include/circt/Dialect/Verif/VerifOps.td

```diff
@@ -260,4 +263,197 @@ def YieldOp : VerifOp<"yield", [
+        verif.ensure %post : i1
+        verif.ensure %post1 : i1
+      }
+      /* ... Module definition ... */
+    }
+    ```
+
+    This later is used to replace any instance of Bar during verification:  
+    ```
+    %bar.0, %bar.1 = hw.instance "bar" @Bar("clk": %clk : !seq.clock, "foo" : %foo : i8)  -> ("" : i8, "1" : i8)
+
+    /* ... After PrepareForFormal Pass becomes ... */
+
+    %c0_8 = hw.constant 0 : i8 
+    %prec = comb.icmp bin neq %c0_8, %bar : i8
+    verif.assert %precond : i1
+
+    %out0 = verif.symbolic_input : i8
+    %out1 = verif.symbolic_input : i8
+    %post = comb.icmp bin ugt %out0, %bar : i8
+    %post1 = comb.icmp bin ult %out1, %bar : i8
+    verif.assume %post : i1
+    verif.assume %post1 : i1
+    ```
+  }];
+  let regions = (region SizedRegion<1>:$contract_region);
+
+  let assemblyFormat = [{
+    attr-dict-with-keyword $contract_region
+  }];
+}
+
+def RequireOp : VerifOp<"require", [
+  HasParent<"verif::ContractOp">, AttrSizedOperandSegments
+]>{
+  let summary = "define a precondition for the given contract";
+  let description = [{
+    This operation defines a precondition for the parent contract. 
+    Preconditions are assumed during the verification of a module and 
+    asserted during the verification of an instance of a module.
+  }];
+  let arguments = (ins LTLAnyPropertyType:$property,
+                      OptionalAttr<ClockEdgeAttr>:$edge, Optional<I1>:$clock,
+                      Optional<I1>:$enable,
+                      OptionalAttr<StrAttr>:$label);
+
+  let assemblyFormat = [{
+    $property (`if` $enable^)? (`,` $edge^ $clock)? 
+    (`label` $label^)? attr-dict `:` type($property)
+  }];
+}
+
+def EnsureOp : VerifOp<"ensure", [
+  HasParent<"verif::ContractOp">, AttrSizedOperandSegments
+]>{
+  let summary = "define a postcondition for the given contract";
+  let description = [{
+    This operation defines a postcondition for the parent contract. 
+    Postconditions are asserted during the verification of a module and 
+    assumed during the verification of an instance of a module.
+  }];
+  let arguments = (ins LTLAnyPropertyType:$property,
+                      OptionalAttr<ClockEdgeAttr>:$edge, Optional<I1>:$clock,
+                      Optional<I1>:$enable,
+                      OptionalAttr<StrAttr>:$label);
+
+  let assemblyFormat = [{
+    $property (`if` $enable^)? (`,` $edge^ $clock)? 
+    (`label` $label^)? attr-dict `:` type($property)
+  }];
+}
+
+def ResultOp : VerifOp<"result", [HasParent<"verif::ResultOp">]>{
+  let summary = "define a symbolic reference to the outputs of the module";
+  let description = [{
+    This operation defines a set of symbolic references to each output of the 
+    module for which this contract is defined. This means that a single reference 
+    of result will result in `#outputs` SSA values which refer to each output of 
+    the module. This is needed for when the instance of the module will be replaced 
```

Comment:
This might be slightly easier to read:
```suggestion
    This operation defines a set of symbolic references to each output of the 
    module within which this contract is defined. This means that a single 
    reference of result will result in as many SSA values as the module has 
    outputs, with each result corresponding to one output. This is needed 
    for when the instance of the module will be replaced 
```